ws281x first compilation on fresh target

simonseres opened the pull request:
Couldn't build it, because of absolute paths to libraries.
This is make output:

`make
g++ -c -pipe -std=c++11 -g -O2 -Wall -W -D_REENTRANT -fPIC -DQT_NO_DEBUG -DQT_PLUGIN -DQT_GUI_LIB -DQT_CORE_LIB -I. -Irpi_ws281x -isystem /usr/include/nymea -isystem /usr/include/arm-linux-gnueabihf/qt5 -isystem /usr/include/arm-linux-gnueabihf/qt5/QtGui -isystem /usr/include/arm-linux-gnueabihf/qt5/QtCore -I. -I/usr/lib/arm-linux-gnueabihf/qt5/mkspecs/linux-g++ -o integrationpluginws281xspi.o integrationpluginws281xspi.cpp
In file included from integrationpluginws281xspi.h:26,
                 from integrationpluginws281xspi.cpp:22:
ws281xspicontroller.h:40:10: fatal error: ../../rpi_ws281x/ws2811.h: No such file or directory
 #include "../../rpi_ws281x/ws2811.h"
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make: *** [Makefile:411: integrationpluginws281xspi.o] Error 1
`

hermanndetz committed:
Initial commit

hermanndetz committed:
Added files missing from previous commit

hermanndetz committed:
Added details about rpi_ws281x lib to README.md

mzanetti commented on debian/control:
```diff
@@ -0,0 +1,27 @@
+               nymea-dev-tools:native,
+               qt5-qmake:native,
+               pkg-config,
+               scons,
```
This fails to install when cross-building. Afaict it should be `scons:native` instead?

mzanetti replied:
Hmm... what does scons actually do?

hermanndetz replied:
scons is the build system of the library that is packaged in the submodule. My initial idea was to keep their build system to avoid side effects if they e.g. split up functionality into different files or something similar. If it causes troubles regarding dependencies, we can as well integrate it into qmake.

hermanndetz replied:
Since the container for crossbuilder is now working, I'm back to this problem. I changed scons to scons:native, which results in unmet dependencies. Couldn't find too much information on using scons within the deb package build system - actually it's discouraged.

```
Using DEB_BUILD_PROFILES: nocheck nodoc noautodbgsym
Building nymea-plugin-ws281xspi for armhf and deploying to device.
LXD container nymea-plugin-ws281xspi-stretch-amd64-armhf already exists.
Error: Common start logic: The container is already running

armhf build dependencies for nymea-plugin-ws281xspi already installed in container nymea-plugin-ws281xspi-stretch-Building nymea-plugin-ws281xspi for armhf in parallel (-j5).
dpkg-buildpackage: info: source package nymea-plugin-ws281xspi
dpkg-buildpackage: info: source version 0.0.1+202010071123~a7d95d7~stretch
dpkg-buildpackage: info: source distribution UNRELEASED
dpkg-buildpackage: info: source changed by  <hermann@base-guh-builder-stretch-amd64-armhf>
dpkg-buildpackage: info: host architecture armhf
 dpkg-source -I -Iobj-* -Idebian/tmp/* -I.bzr* --before-build nymea-plugin-ws281xspi
dpkg-checkbuilddeps: error: Unmet build dependencies: scons:native
dpkg-buildpackage: warning: build dependencies/conflicts unsatisfied; aborting
dpkg-buildpackage: warning: (Use -d flag to override.)

```

So the next best guess would be to include the files into the plugin's project file. Yet this would rely on the assumption that the file structure does not change in future versions.

Or as another option to play it safe, include the source code files of the library in their current version into the repo here.

mzanetti replied:
Well, when using a git submodule, the used commit hash is stored along with it. So any checkout would always use the same revision. Unless the project devs force-push to their master branch and break the history, adding the files to the .pro in here seems to be fine IMO. If you see the risk of them breaking their master branch history, we might as well fork it on github and use our fork as the submodule.

simonseres commented:
when initialising and updating the git submodule, the rpi_ws281x library folder is within the nymea-plugin-ws281x folder, but the code/makefile expect it to be next to it.

simonseres commented:
Please add in the readme that scons is required: `apt install scons`

simonseres commented:
Adding the plugin leads to a segvault on a standard nymea image.
`Cannot open /dev/spidev0.0. spi_bcm2835 module not loaded?`
Please describe in README how to enable SPI using `raspi-config` as a prerequisite.

mzanetti commented:
I fail to build it with crossbuilder... Not yet sure why... The error message is not very helpful. dh_auto_configure returned status 10... Any ideas?

simonseres commented:
> I fail to build it with crossbuilder... Not yet sure why... The error message is not very helpful. dh_auto_configure returned status 10... Any ideas?

The 2 issues I had to solve: 
1) copy the rpi_ws281x library files to a folder underneath the actual plugin folder containing the .pro file to make the build process work through. https://github.com/nymea/nymea-plugin-ws281x/pull/1#issuecomment-688465972
2) install scons 

Maybe one of those issues is related to the crossbuild error

mzanetti commented on plugininfo.h:
```diff
@@ -0,0 +1,100 @@
+/* This file is generated by the nymea build system. Any changes to this file will *
+ * be lost. If you want to change this file, edit the plugin's json file.          */
+
+#ifndef PLUGININFO_H
```
You shouldn't commit this file. it'll be autogenerated on every build anyways.

hermanndetz replied:
Sorry, my mistake. I actually know that it's generated upon build. I guess the best fix for this would be to add it to .gitignore?

mzanetti replied:
The thing with gitignore is that it breaks `git clean`...

mzanetti replied:
Also, the best fix is to do out-of source builds :)

mzanetti commented on ws281xspi.pro:
```diff
@@ -0,0 +1,17 @@
+include($$[QT_INSTALL_PREFIX]/include/nymea/plugin.pri)
+
+QMAKE_PRE_LINK = scons -C $$_PRO_FILE_PWD_/rpi_ws281x libws2811.a
```
Doesn't look like this actually does what you want, does it?

afaict the plugin is now built first with make, and then you're running scons on it again just to print that make already did everything...  Maybe I'm wrong... never used scons so far.

hermanndetz replied:
No, this file actually builds the library so it can be linked with the plugin in the next step. If the conclusion is to drop scons, this line would of course disappear also.

mzanetti replied:
```
14:03:04: Running steps for project ws281xspi...
14:03:04: Starting: "/usr/lib/qt5/bin/qmake" /home/micha/Develop/nymea-plugin-ws281x/ws281xspi.pro -spec linux-g++ CONFIG+=debug CONFIG+=qml_debug PLUGIN_PRI=/home/micha/Develop/nymea/libnymea/integrations/plugin.pri
Project MESSAGE: ****
14:03:04: The process "/usr/lib/qt5/bin/qmake" exited normally.
14:03:04: Starting: "/usr/bin/make" -f /home/micha/Develop/build-ws281xspi-Desktop-Debug/Makefile qmake_all
make: Nothing to be done for 'qmake_all'.
14:03:04: The process "/usr/bin/make" exited normally.
14:03:04: Starting: "/usr/bin/make" -j8
g++ -c -pipe -Werror -std=c++11 -g -g -Wall -W -D_REENTRANT -fPIC -DQT_QML_DEBUG -DQT_PLUGIN -DQT_GUI_LIB -DQT_CORE_LIB -I../nymea-plugin-ws281x -I. -I../nymea-plugin-ws281x/rpi_ws281x -isystem /usr/include/nymea/ -isystem /usr/include/x86_64-linux-gnu/qt5 -isystem /usr/include/x86_64-linux-gnu/qt5/QtGui -isystem /usr/include/x86_64-linux-gnu/qt5/QtCore -I. -I/usr/lib/x86_64-linux-gnu/qt5/mkspecs/linux-g++ -o integrationpluginws281xspi.o ../nymea-plugin-ws281x/integrationpluginws281xspi.cpp
nymea-plugininfocompiler ../nymea-plugin-ws281x/integrationpluginws281xspi.json --extern extern-plugininfo.h
g++ -pipe -Werror -std=c++11 -g -g -Wall -W -dM -E -o moc_predefs.h /usr/lib/x86_64-linux-gnu/qt5/mkspecs/features/data/dummy.cpp
/usr/lib/qt5/bin/moc -DQT_QML_DEBUG -DQT_PLUGIN -DQT_GUI_LIB -DQT_CORE_LIB --include /home/micha/Develop/build-ws281xspi-Desktop-Debug/moc_predefs.h -I/usr/lib/x86_64-linux-gnu/qt5/mkspecs/linux-g++ -I/home/micha/Develop/nymea-plugin-ws281x -I/home/micha/Develop/nymea-plugin-ws281x/rpi_ws281x -I/usr/include/nymea -I/usr/include/x86_64-linux-gnu/qt5 -I/usr/include/x86_64-linux-gnu/qt5/QtGui -I/usr/include/x86_64-linux-gnu/qt5/QtCore -I. -I/home/micha/Develop/nymea/libnymea -I/usr/include/c++/9 -I/usr/include/x86_64-linux-gnu/c++/9 -I/usr/include/c++/9/backward -I/usr/lib/gcc/x86_64-linux-gnu/9/include -I/usr/local/include -I/usr/include/x86_64-linux-gnu -I/usr/include ../nymea-plugin-ws281x/integrationpluginws281xspi.h -o moc_integrationpluginws281xspi.cpp
/usr/lib/qt5/bin/moc -DQT_QML_DEBUG -DQT_PLUGIN -DQT_GUI_LIB -DQT_CORE_LIB --include /home/micha/Develop/build-ws281xspi-Desktop-Debug/moc_predefs.h -I/usr/lib/x86_64-linux-gnu/qt5/mkspecs/linux-g++ -I/home/micha/Develop/nymea-plugin-ws281x -I/home/micha/Develop/nymea-plugin-ws281x/rpi_ws281x -I/usr/include/nymea -I/usr/include/x86_64-linux-gnu/qt5 -I/usr/include/x86_64-linux-gnu/qt5/QtGui -I/usr/include/x86_64-linux-gnu/qt5/QtCore -I. -I/home/micha/Develop/nymea/libnymea -I/usr/include/c++/9 -I/usr/include/x86_64-linux-gnu/c++/9 -I/usr/include/c++/9/backward -I/usr/lib/gcc/x86_64-linux-gnu/9/include -I/usr/local/include -I/usr/include/x86_64-linux-gnu -I/usr/include ../nymea-plugin-ws281x/ws281xspicontroller.h -o moc_ws281xspicontroller.cpp
g++ -c -pipe -Werror -std=c++11 -g -g -Wall -W -D_REENTRANT -fPIC -DQT_QML_DEBUG -DQT_PLUGIN -DQT_GUI_LIB -DQT_CORE_LIB -I../nymea-plugin-ws281x -I. -I../nymea-plugin-ws281x/rpi_ws281x -isystem /usr/include/nymea/ -isystem /usr/include/x86_64-linux-gnu/qt5 -isystem /usr/include/x86_64-linux-gnu/qt5/QtGui -isystem /usr/include/x86_64-linux-gnu/qt5/QtCore -I. -I/usr/lib/x86_64-linux-gnu/qt5/mkspecs/linux-g++ -o ws281xspicontroller.o ../nymea-plugin-ws281x/ws281xspicontroller.cpp
g++ -c -pipe -Werror -std=c++11 -g -g -Wall -W -D_REENTRANT -fPIC -DQT_QML_DEBUG -DQT_PLUGIN -DQT_GUI_LIB -DQT_CORE_LIB -I../nymea-plugin-ws281x -I. -I../nymea-plugin-ws281x/rpi_ws281x -isystem /usr/include/nymea/ -isystem /usr/include/x86_64-linux-gnu/qt5 -isystem /usr/include/x86_64-linux-gnu/qt5/QtGui -isystem /usr/include/x86_64-linux-gnu/qt5/QtCore -I. -I/usr/lib/x86_64-linux-gnu/qt5/mkspecs/linux-g++ -o moc_ws281xspicontroller.o moc_ws281xspicontroller.cpp
g++ -c -pipe -Werror -std=c++11 -g -g -Wall -W -D_REENTRANT -fPIC -DQT_QML_DEBUG -DQT_PLUGIN -DQT_GUI_LIB -DQT_CORE_LIB -I../nymea-plugin-ws281x -I. -I../nymea-plugin-ws281x/rpi_ws281x -isystem /usr/include/nymea/ -isystem /usr/include/x86_64-linux-gnu/qt5 -isystem /usr/include/x86_64-linux-gnu/qt5/QtGui -isystem /usr/include/x86_64-linux-gnu/qt5/QtCore -I. -I/usr/lib/x86_64-linux-gnu/qt5/mkspecs/linux-g++ -o moc_integrationpluginws281xspi.o moc_integrationpluginws281xspi.cpp
scons -C /home/micha/Develop/nymea-plugin-ws281x/rpi_ws281x libws2811.a
scons: Entering directory `/home/micha/Develop/nymea-plugin-ws281x/rpi_ws281x'
scons: Reading SConscript files ...
scons: done reading SConscript files.
scons: Building targets ...
scons: `libws2811.a' is up to date.
scons: done building targets.
rm -f libnymea_integrationpluginws281xspi.so
g++ -std=c++11 -shared -o libnymea_integrationpluginws281xspi.so integrationpluginws281xspi.o ws281xspicontroller.o moc_integrationpluginws281xspi.o moc_ws281xspicontroller.o  -L/home/micha/Develop/nymea-plugin-ws281x/rpi_ws281x/ -lws2811 -lnymea /usr/lib/x86_64-linux-gnu/libQt5Gui.so /usr/lib/x86_64-linux-gnu/libQt5Core.so /usr/lib/x86_64-linux-gnu/libGL.so -lpthread    
14:03:06: The process "/usr/bin/make" exited normally.
14:03:06: Elapsed time: 00:02.
```

To me this really looks like it's running make and when make is done it runs scons which just says "scons: `libws2811.a' is up to date."

mzanetti replied:
Oh... stupid me.. this just builds the submodule stuff... Right...

simonseres commented:
Explain in README what Pin / GPIO you mean exactly by "Pin 10".
Sadly there is some complexity in naming the pins on the pi.
Maybe you can help users by highlighting the right Pin in the Raspberry Pi header diagram.

<img width="798" alt="image" src="https://user-images.githubusercontent.com/6623401/92412732-49229380-f14d-11ea-8b08-29c25d59258c.png">

hermanndetz committed:
Removed plugininfo.h

This file is generated by nymea-plugininfocompiler.

hermanndetz committed:
Extended README.md

Added section on hardware connection (pins + activation using
raspi-config)

hermanndetz committed:
Added note about scons to README.md

hermanndetz committed:
Fixed crossbuilder build

Removed scons dependency
Fixed include path for lib in submodule
Added translations file
